src/main/scala/org/constellation/domain/redownload/RedownloadService.scala

```diff
@@ -1,38 +1,130 @@
 package org.constellation.domain.redownload
 
+import cats.effect.{ContextShift, LiftIO, Sync}
 import cats.effect.Concurrent
 import cats.effect.concurrent.Ref
 import cats.implicits._
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+import org.constellation.DAO
+import org.constellation.consensus.EdgeProcessor
+import org.constellation.domain.snapshotInfo.SnapshotInfoChunk
 import org.constellation.schema.Id
+import org.constellation.storage.RecentSnapshot
+import org.constellation.util.MajorityStateChooser.NodeSnapshots
+import org.constellation.util.{HealthChecker, MajorityStateChooser}
 
-class RedownloadService[F[_]]()(implicit F: Concurrent[F]) {
+import scala.concurrent.duration._
+
+class RedownloadService[F[_]](dao: DAO)(implicit F: Concurrent[F], C: ContextShift[F]) {
+  val logger = Slf4jLogger.getLogger[F]
 
   // TODO: Consider Height/Hash type classes
-  private[redownload] val ownSnapshots: Ref[F, Map[Long, String]] = Ref.unsafe(Map.empty)
-  private[redownload] val peersProposals: Ref[F, Map[Long, (Id, String)]] = Ref.unsafe(Map.empty)
+  private[redownload] val ownSnapshots: Ref[F, Map[Long, RecentSnapshot]] = Ref.unsafe(Map.empty)
+  private[redownload] val peersProposals
+    : Ref[F, Map[Long, Seq[NodeSnapshots]]] = Ref.unsafe(Map.empty) //todo join/leave?
```

Comment:
IMO, it would be way easier to keep proposals as `Map[Id, Proposals]` where `type Proposals = Map[Long, String]`. To update it, it's just a matter of `existingProposals ++ newProposals` for each node id

Reply:
Yeah we might as well